File: java/revenj-storage/src/main/java/org/revenj/AmazonS3Repository.java
package org.revenj;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import java.util.Optional;
import java.util.Properties;
import java.util.concurrent.*;

import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.regions.Region;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.s3.AmazonS3Client;
import com.amazonaws.services.s3.model.DeleteObjectRequest;
import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import org.revenj.storage.S3Repository;

class AmazonS3Repository implements S3Repository, Closeable {
	private final String bucketName;
	private final ExecutorService executorService;
	private final AmazonS3Client s3Client;
	private final boolean disposeExecutor;

	public AmazonS3Repository(Properties properties, Optional<ExecutorService> executorService) {
		bucketName = properties.getProperty("revenj.s3-bucket-name");
		String s3AccessKey = properties.getProperty("revenj.s3-user");
		String s3SecretKey = properties.getProperty("revenj.s3-secret");
		String s3Region = properties.getProperty("revenj.s3-region");
		disposeExecutor = !executorService.isPresent();
		this.executorService = executorService.orElse(Executors.newSingleThreadExecutor());
		if (s3AccessKey == null || s3AccessKey.isEmpty()) {
			throw new RuntimeException("S3 configuration is missing. Please add revenj.s3-user");
		}
		if (s3SecretKey == null || s3SecretKey.isEmpty()) {
			throw new RuntimeException("S3 configuration is missing. Please add revenj.s3-secret");
		}
		s3Client = new AmazonS3Client(new BasicAWSCredentials(s3AccessKey, s3SecretKey));
		if (s3Region != null) {
			s3Client.setRegion(Region.getRegion(Regions.fromName(s3Region)));
		}
	}

	private String getBucketName(final String name) throws IOException {
		String bn = name == null || name.isEmpty() ? bucketName : name;
		if (bn == null || bn.isEmpty()) {
			throw new IOException("Bucket name not specified for this S3 instance or system wide.\n"
					+ "Either specify revenj.s3-bucket-name in Properties as system wide name or provide a bucket name to this S3 instance");
		}
		return bn;
	}

	@Override
	public Future<InputStream> get(final String bucket, final String key) {
		return executorService.submit(() -> {
			S3Object s3 = s3Client.getObject(new GetObjectRequest(bucket, key));
			return s3.getObjectContent();
		});
	}

	@Override
	public Future<Void> upload(
			String bucket,
			String key,
			InputStream stream,
			long length,
			Map<String, String> metadata) {
		return executorService.submit(() -> {
			ObjectMetadata om = new ObjectMetadata();
			om.setContentLength(length);
			if (metadata != null) {
				for (final Map.Entry<String, String> kv : metadata.entrySet()) {
					om.addUserMetadata(kv.getKey(), kv.getValue());
				}
			}
			s3Client.putObject(new PutObjectRequest(getBucketName(bucket), key, stream, om));
			return null;
		});
	}

	@Override
	public Future<Void> delete(String bucket, String key) {
		return executorService.submit(() -> {
			s3Client.deleteObject(new DeleteObjectRequest(getBucketName(bucket), key));
			return null;
		});
	}

	@Override
	public void close() throws IOException {
		if (disposeExecutor) {
			executorService.shutdown();
		}
	}
}

File: scala/revenj-storage/src/main/scala/net/revenj/AmazonS3Repository.scala
package net.revenj

import net.revenj.storage.{S3, S3Repository}
import software.amazon.awssdk.auth.credentials.{AwsBasicCredentials, StaticCredentialsProvider}
import software.amazon.awssdk.core.sync.RequestBody
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.s3.S3Client
import software.amazon.awssdk.services.s3.model._

import java.io.InputStream
import java.util.{Properties, UUID}
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._

class AmazonS3Repository(
  properties: Properties,
  tryExecutionContext: Option[ExecutionContext],
  tryS3: Option[S3Client]
) extends S3Repository {

  private val executionContext = tryExecutionContext.getOrElse(ExecutionContext.global)
  private val s3Client = tryS3.getOrElse{
    val s3AccessKey = Option(properties.getProperty("revenj.s3-user"))
    val s3SecretKey = Option(properties.getProperty("revenj.s3-secret"))
    val s3Region = Option(properties.getProperty("revenj.s3-region"))
    require(s3AccessKey.isDefined && s3AccessKey.get.nonEmpty, "S3 configuration is missing. Please add revenj.s3-user")
    require(s3SecretKey.isDefined && s3SecretKey.get.nonEmpty, "S3 configuration is missing. Please add revenj.s3-secret")
    val builder = S3Client.builder()
    builder.credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create(s3AccessKey.get, s3SecretKey.get)))
    s3Region.foreach(r => builder.region(Region.of(r)))
    builder.build()
  }

  private val bucketName = Option(properties.getProperty("revenj.s3-bucket-name"))
  private def getBucketName(name: String) = {
    val bn = if (name == null || name.isEmpty) bucketName else Some(name)
    require(bn.isDefined && bn.get.nonEmpty, """Bucket name not specified for this S3 instance or system wide.
Either specify revenj.s3-bucket-name in Properties as system wide name or provide a bucket name to this S3 instance""")
    bn.get
  }

  override def get(bucket: String, key: String): Future[InputStream] = {
    Future {
      s3Client.getObject(GetObjectRequest.builder()
        .bucket(bucket)
        .key(key)
        .build()
      )
    }(executionContext)
  }

  override def upload(stream: InputStream, length: Long): Future[S3] = {
    val bn = getBucketName("")
    val key = UUID.randomUUID.toString
    val putObjectRequest = PutObjectRequest.builder()
      .bucket(bn)
      .key(key)
      .contentLength(length)
      .build()
    Future {
      s3Client.putObject(putObjectRequest, RequestBody.fromInputStream(stream, length))
      S3(bn, key, length)
    }(executionContext)
  }

  override def upload(
    bucket: String,
    key: String,
    stream: InputStream,
    length: Long,
    name: Option[String],
    mimeType: Option[String],
    metadata: Map[String, String],
    tags: Map[String, String]
  ): Future[S3] = {
    val bn = getBucketName(bucket)
    val javaTags = tags.map { case (key, value) =>
      Tag.builder().key(key).value(value).build()
    }.toSet.asJava
    val putObjectRequestBuilder = PutObjectRequest.builder()
      .bucket(bn)
      .key(key)
      .contentLength(length)
      .metadata(metadata.asJava)
      .tagging(Tagging.builder().tagSet(javaTags).build())
    mimeType.foreach(putObjectRequestBuilder.contentType)
    val putObjectRequest = putObjectRequestBuilder.build()
    Future {
      s3Client.putObject(putObjectRequest, RequestBody.fromInputStream(stream, length))
      S3(bn, key, length, name, mimeType, metadata)
    }(executionContext)
  }

  override def delete(bucket: String, key: String): Future[DeleteObjectResponse] = {
    val deleteObjectRequest = DeleteObjectRequest.builder()
      .bucket(bucket)
      .key(key)
      .build()
    Future {
      s3Client.deleteObject(deleteObjectRequest)
    }(executionContext)
  }
}
